src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+          }
+
+          blockData[prototypeId] = {
+            opcode: OpCode.procedures_prototype,
+
+            next: null,
+            parent: block.id,
+
+            inputs: prototypeInputs,
+            fields: {},
+
+            shadow: true,
+            topLevel: false,
+
+            mutation: {
+              tagName: "mutation",
+              children: [],
+              proccode: block.inputs.PROCCODE.value,
+              argumentids: JSON.stringify(args.map(arg => arg.id)),
+              argumentnames: JSON.stringify(args.map(arg => arg.name)),
+              argumentdefaults: JSON.stringify(args.map(arg => arg.default)),
+              warp: JSON.stringify(warp) as "true" | "false"
+            }
+          };
+
+          inputs.custom_block = [BIS.INPUT_SAME_BLOCK_SHADOW, prototypeId];
+
+          break;
+        }
+
+        case OpCode.procedures_call: {
+          const proccode = block.inputs.PROCCODE.value;
+          const customBlockData = getCustomBlockData(proccode);
+          if (!customBlockData) {
+            warn(
+              `Missing custom block prototype for proccode ${proccode} (${block.id} in ${target.name}); skipping this block`
+            );
+            return null;
+          }
+
+          const { args, warp } = customBlockData;
+
+          mutation = {
+            tagName: "mutation",
+            children: [],
+            proccode: block.inputs.PROCCODE.value,
+            argumentids: JSON.stringify(args.map(arg => arg.id)),
+            warp: JSON.stringify(warp) as "true" | "false"
+          };
+
+          const inputEntries = {};
+          const constructedInputs = {};
+          const initialValues = {};
+          let i = 0;
+          for (const arg of args) {
+            inputEntries[arg.id] = prop(
+              {
+                boolean: sb3.BooleanOrSubstackInputStatus,
+                numberOrString: BIS.TEXT_PRIMITIVE
+              },
+              arg.type
+            );
+            initialValues[arg.id] = prop(
+              {
+                boolean: false,
+                numberOrString: ""
+              },
+              arg.type
+            );
+            constructedInputs[arg.id] = block.inputs.INPUTS.value[i++];
+          }
+
+          const result = serializeInputsToInputs(constructedInputs, {
+            stage,
+            target,
+
+            getBroadcastId,
+            getCustomBlockData,
+
+            block,
+            initialValues,
+            inputEntries
+          });
+
+          inputs = result.inputs;
+          applyBlockData(blockData, result.blockData);
+
+          break;
+        }
+
+        default: {
+          const inputEntries = prop(sb3.inputPrimitiveOrShadowMap, block.opcode);
+
+          const initialValues = {};
+          for (const key of Object.keys(inputEntries)) {
+            initialValues[key] = (BlockBase.getDefaultInput(block.opcode, key) || {}).initial;
```

Comment:
```suggestion
            const defaultInput = BlockBase.getDefaultInput(block.opcode, key);
            if (defaultInput)) {
              initialValues[key] = defaultInput.initial;
            }
```

Not sure if this changes the behavior since setting object values to `undefined` is not the same as never adding them in the first place, but `JSON.stringify` doesn't serialize `undefined` properties anyway.